test/unittests/t_payload_processor.cc

```diff
@@ -0,0 +1,102 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include <gtest/gtest.h>
+
+#include "pack.h"
+#include "receiver/payload_processor.h"
+#include "util/string.h"
+#include "util_concurrency.h"
+
+using namespace receiver;  // NOLINT
+
+class MockPayloadProcessor : public PayloadProcessor {
+ public:
+  MockPayloadProcessor() : num_files_received_(0) {}
+  virtual ~MockPayloadProcessor() {}
+
+  virtual void ConsumerEventCallback(const ObjectPackBuild::Event& /*event*/) {
+    num_files_received_++;
+  }
+
+  int num_files_received_;
+};
+
+class T_PayloadProcessor : public ::testing::Test {
+ protected:
+  T_PayloadProcessor()
+      : ready_(1, 1), serializer_(NULL), digest_(shash::kSha1) {}
+
+  virtual void SetUp() {
+    int fds[2];
+    ASSERT_NE(pipe(fds), -1);
+    read_fd_ = fds[0];
+
+    // Prepare an object pack and send it through the pipe
+    ObjectPack::BucketHandle hd = pack_.NewBucket();
+
+    std::vector<uint8_t> buffer(4096, 0);
+    ObjectPack::AddToBucket(&buffer[0], 4096, hd);
+    shash::Any buffer_hash(shash::kSha1);
+    shash::HashMem(&buffer[0], buffer.size(), &buffer_hash);
+    ASSERT_TRUE(pack_.CommitBucket(ObjectPack::kCas, buffer_hash, hd));
+
+    serializer_ = new ObjectPackProducer(&pack_);
+
+    ASSERT_TRUE(serializer_);
+
+    serializer_->GetDigest(&digest_);
+
+    write_fd_ = fds[1];
+
+    ASSERT_EQ(pthread_create(&thread_, NULL, T_PayloadProcessor::Worker,
+                             static_cast<void*>(this)),
+              0);
+
+    ASSERT_EQ(ready_.Dequeue(), true);
+  }
+
+  virtual void TearDown() {
+    ASSERT_EQ(pthread_join(thread_, NULL), 0);
```

Comment:
Unfortunately, GTest would like to have arguments in `ASSERT_EQ` and `EXPECT_EQ` in reverse order for proper error reporting, like `ASSERT_EQ(0, pthread_join(thread_, NULL));`